sdk/messaging/azservicebus/internal/namespace.go

```diff
@@ -303,62 +308,74 @@ func (ns *Namespace) Recover(ctx context.Context) (*amqp.Client, error) {
 // negotiateClaim performs initial authentication and starts periodic refresh of credentials.
 // the returned func is to cancel() the refresh goroutine.
 func (ns *Namespace) NegotiateClaim(ctx context.Context, entityPath string) (func() <-chan struct{}, error) {
-	return ns.negotiateClaimImpl(ctx,
+	return ns.startNegotiateClaimRenewer(ctx,
 		entityPath,
-		15*time.Minute,
-		func(ctx context.Context, audience string, conn *amqp.Client, provider auth.TokenProvider) error {
-			// You're not allowed to have multiple $cbs links open in a single connect. The current
-			// cbs.NegotiateClaim implementation automatically creates and shuts down it's own
-			// link so we have to guard against that here (for now).
-			ns.negotiateClaimMu.Lock()
-			defer ns.negotiateClaimMu.Unlock()
-			return cbs.NegotiateClaim(ctx, audience, conn, provider)
-		},
-		ns.getAMQPClientImpl,
-		func(err error) {
-			tab.For(ctx).Debug(fmt.Sprintf("refreshing claims failed: %s", err.Error()))
-		})
+		cbs.NegotiateClaim,
+		ns.getAMQPClientImpl)
 }
 
-func (ns *Namespace) negotiateClaimImpl(ctx context.Context,
+func (ns *Namespace) startNegotiateClaimRenewer(ctx context.Context,
 	entityPath string,
-	renewalDuration time.Duration,
 	cbsNegotiateClaim func(ctx context.Context, audience string, conn *amqp.Client, provider auth.TokenProvider) error,
-	nsGetAMQPClientImpl func(ctx context.Context) (*amqp.Client, error),
-	logError func(err error),
-) (func() <-chan struct{}, error) {
+	nsGetAMQPClientImpl func(ctx context.Context) (*amqp.Client, error)) (func() <-chan struct{}, error) {
 	audience := ns.GetEntityAudience(entityPath)
 
-	refreshClaim := func() error {
-		amqpClient, err := nsGetAMQPClientImpl(ctx)
-
-		if err != nil {
-			logError(err)
-			return err
-		}
-
+	refreshClaim := func() (time.Time, error) {
 		retrier := ns.baseRetrier.Copy()
 
+		var lastErr error
+		var expiration time.Time
+
 		for retrier.Try(ctx) {
-			ctx, span := ns.startSpanFromContext(ctx, "sb.namespace.negotiateClaim")
-			err = cbsNegotiateClaim(ctx, audience, amqpClient, ns.TokenProvider)
+			expiration, lastErr = func() (time.Time, error) {
+				ctx, span := ns.startSpanFromContext(ctx, "sb.namespace.negotiateClaim")
+				defer span.End()
+
+				amqpClient, err := nsGetAMQPClientImpl(ctx)
+
+				if err != nil {
+					span.Logger().Error(err)
+					return time.Time{}, err
+				}
+
+				token, expiration, err := ns.TokenProvider.GetTokenAsTokenProvider(audience)
+
+				if err != nil {
```

Comment:
What happens when refreshing the token fails when the token hasn't yet expired?

Reply:
I rebased so this might have moved your comment to the wrong line.

I think you're asking me whether the `expiration` there is used if there is a failure. We don't use it - we only use if it we succeeded.

There is a retry loop around the negotiate itself, so we do have some fallback logic there. There is also a "pause 1 minute and then try 5 times, per 5 seconds" check later in the function.

Reply:
Ah, offline discussion clarified the question somewhat.

Service Bus authenticates differently than other HTTP stacks in that the token is not passed for each link's request, but instead done as a whole for the connection, out of band from the individual requests. It's described here if you're interested:

https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-protocol-guide#claims-based-authorization

So essentially this code is just requesting/updating a token from the TokenProvider and notifying Service Bus about it. All previous or current tokens are just local to this loop, there's no stored token state in our code.